File: src/flaskapp/static/js/runner.js
var runner_timeout = 10000;
var runner_polling_interval = 512;

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function fetchJson(url,data){
    return await $.ajax({
        crossDomain: true,
        url: url,
        type: "POST",
        data: data,
    });
}

async function runner_api_test(){
    var a = await runnerGetDetails("db_QO3lJlolQ8TQ2SzL1ng")
    console.log(a)
}

function runnerCreate(scriptLang, scriptContents, scriptInput){
    var params = {
        source_code: scriptContents,
        language: scriptLang,
        input: scriptInput
    };
    return fetchJson('/api/runner/create',params);
}
function runnerGetStatus(paramid){
    var params = {
        id:paramid
    };
    return fetchJson('/api/runner/get_status',params);
}
function runnerGetDetails(paramid){
    var params = {
        id:paramid
    };
    return fetchJson('/api/runner/get_details',params);
}

async function runScript(scriptLang, scriptContents, scriptInput){
    var runner_session = await runnerCreate(scriptLang,scriptContents,scriptInput);
    var status = runner_session.status;
    var timeout_left = runner_timeout;

    while(status!='completed' && timeout_left>0){
        await sleep(runner_polling_interval);
        timeout_left -= runner_polling_interval;
        var result = await runnerGetStatus(runner_session.id);
        status = result.status;
    };
    if(status!='completed')
        return {
            result:'failure',
            status:'completed',
            stderr:'timeout'
        }
    else
        return await runnerGetDetails(runner_session.id);
}

File: src/flaskapp/runner.py
import requests
from flask import jsonify

_urls={
    'create':'http://api.paiza.io:80/runners/create',
    'get_status':'http://api.paiza.io:80/runners/get_status',
    'get_details':'http://api.paiza.io:80/runners/get_details'
}
_api_key = 'guest'

def _fetch_json_post(url, params):
    result = requests.post(url,json=params)
    return jsonify(result.json())

def _fetch_json(url, params):
    result = requests.get(url,params=params)
    return jsonify(result.json())

def create(scriptLang, scriptContents, scriptInput):
    params = {
        'source_code': scriptContents,
        'language': scriptLang,
        'input':scriptInput,
        'api_key':_api_key
    }
    return _fetch_json_post(_urls['create'], params)


def get_status(session_id):
    params = {
        'id': session_id,
        'api_key':_api_key
    }
    return _fetch_json(_urls['get_status'], params)


def get_details(session_id):
    params = {
        'id': session_id,
        'api_key':_api_key
    }
    return _fetch_json(_urls['get_details'], params)
